From c490ead3844de6dda92fba995df975c6c100086f Mon Sep 17 00:00:00 2001
From: Hanna Hawa <hannah@marvell.com>
Date: Wed, 21 Jan 2015 21:17:44 +0200
Subject: [PATCH 0103/1240] fix: mpp: enable MPP configuration for Armada8k

	- Add fix for Armada-8k, the MPP high register at low address
	- Fix the base address of A8k
	- Add MPP values for A8k boards

Signed-off-by: Hanna Hawa <hannah@marvell.com>
Change-Id: I7d57489483591444d84c2ef8589809d360abbe17
Reviewed-on: http://vgitil04.il.marvell.com:8080/16180
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
---
 arch/arm/cpu/armv8/armada8k/soc-init.c         | 2 +-
 arch/arm/cpu/mvebu-common/mpp.c                | 5 ++++-
 arch/arm/include/asm/arch-armada8k/regs-base.h | 2 +-
 arch/arm/include/asm/arch-armada8k/soc-info.h  | 2 +-
 board/mvebu/armada8k/db-8210-info.c            | 1 +
 board/mvebu/armada8k/rd-8210-info.c            | 1 +
 board/mvebu/common/init.c                      | 3 ---
 include/configs/mvebu-common.h                 | 1 +
 8 files changed, 10 insertions(+), 7 deletions(-)

diff --git a/arch/arm/cpu/armv8/armada8k/soc-init.c b/arch/arm/cpu/armv8/armada8k/soc-init.c
index 89ba703..8e180ab 100644
--- a/arch/arm/cpu/armv8/armada8k/soc-init.c
+++ b/arch/arm/cpu/armv8/armada8k/soc-init.c
@@ -39,7 +39,7 @@ int soc_get_rev(void)
 int soc_get_id(void)
 {
 	/* This should read the soc id from some register*/
-	return 8022;
+	return CONFIG_ARMADA_8K_SOC_ID;
 }
 
 struct mvebu_soc_family *soc_init(void)
diff --git a/arch/arm/cpu/mvebu-common/mpp.c b/arch/arm/cpu/mvebu-common/mpp.c
index dcd615c..a1e883a 100644
--- a/arch/arm/cpu/mvebu-common/mpp.c
+++ b/arch/arm/cpu/mvebu-common/mpp.c
@@ -120,7 +120,10 @@ void mpp_set_and_update(u32 *mpp_reg)
 		debug("Set mpp reg 0x%08x\n", mpp_reg[i]);
 
 		/* Write to register */
-		writel(mpp_reg[i], base + i);
+		if (soc_get_id() == CONFIG_ARMADA_8K_SOC_ID)
+			writel(mpp_reg[i], base - i);
+		else
+			writel(mpp_reg[i], base + i);
 	}
 }
 
diff --git a/arch/arm/include/asm/arch-armada8k/regs-base.h b/arch/arm/include/asm/arch-armada8k/regs-base.h
index aac7e01..95e2b32 100644
--- a/arch/arm/include/asm/arch-armada8k/regs-base.h
+++ b/arch/arm/include/asm/arch-armada8k/regs-base.h
@@ -22,7 +22,7 @@
 
 #include <asm/arch/memory-map.h>
 
-#define MPP_REGS_BASE	(MVEBU_REGS_BASE + 0x1000)
+#define MPP_REGS_BASE	(MVEBU_REGS_BASE + 0x6F008C)
 
 /* List of register base for all units */
 #define MVEBU_ADEC_AP_BASE 	(MVEBU_REGS_BASE + 0x0)
diff --git a/arch/arm/include/asm/arch-armada8k/soc-info.h b/arch/arm/include/asm/arch-armada8k/soc-info.h
index a670951..7b4cb4a 100644
--- a/arch/arm/include/asm/arch-armada8k/soc-info.h
+++ b/arch/arm/include/asm/arch-armada8k/soc-info.h
@@ -22,7 +22,7 @@
 
 /* General MPP definitions */
 #define MAX_MPP_OPTS		7
-#define MAX_MPP_ID		69
+#define MAX_MPP_ID		15
 
 #define MPP_BIT_CNT		4
 #define MPP_FIELD_MASK		0x7
diff --git a/board/mvebu/armada8k/db-8210-info.c b/board/mvebu/armada8k/db-8210-info.c
index 1b005e0..e5626d8 100644
--- a/board/mvebu/armada8k/db-8210-info.c
+++ b/board/mvebu/armada8k/db-8210-info.c
@@ -32,6 +32,7 @@ struct mvebu_board_config db_a8021_config = {
 struct mvebu_board_info a8021_db_info = {
 	.name = "DB-MV8021",
 	.id = ARMADA_8021_DB_ID,
+	.mpp_regs = {0x11111111, 0x00100001},
 	.configurable = 1,
 	.config_data = &db_a8021_config,
 };
diff --git a/board/mvebu/armada8k/rd-8210-info.c b/board/mvebu/armada8k/rd-8210-info.c
index deb2a23..f934a56 100644
--- a/board/mvebu/armada8k/rd-8210-info.c
+++ b/board/mvebu/armada8k/rd-8210-info.c
@@ -38,6 +38,7 @@ u16 a8021_rd_unit_disable[MAX_UNIT_ID] = {
 struct mvebu_board_info a8021_rd_info = {
 	.name = "RD-MV8021",
 	.id = ARMADA_8021_RD_ID,
+	.mpp_regs = {0x11111111, 0x00100001},
 	.unit_mask = a8021_rd_unit_disable,
 	.unit_update_mode = UNIT_INFO_DISABLE,
 	.configurable = 1,
diff --git a/board/mvebu/common/init.c b/board/mvebu/common/init.c
index 4bf2f96..c4b5104 100644
--- a/board/mvebu/common/init.c
+++ b/board/mvebu/common/init.c
@@ -142,11 +142,8 @@ int mvebu_board_init(void)
 	mvebu_devel_board_init(brd_fam);
 #endif
 
-#ifndef CONFIG_PALLADIUM
 	/* mpp_set */
 	mpp_set_and_update(brd->mpp_regs);
-#endif
-
 
 	return 0;
 }
diff --git a/include/configs/mvebu-common.h b/include/configs/mvebu-common.h
index 881415a..1d2eaf4 100644
--- a/include/configs/mvebu-common.h
+++ b/include/configs/mvebu-common.h
@@ -34,6 +34,7 @@
 /* #define CONFIG_MV_DEBUG_PUTS */
 
 #define CONFIG_SYS_GENERIC_BOARD
+#define CONFIG_ARMADA_8K_SOC_ID		8022
 
 /* Global definitions */
 #ifndef CONFIG_SYS_TEXT_BASE
-- 
1.9.1

